### PR TITLE
Implement mask for some primitives + jit.

### PR DESCRIPTION
**Pull request by juliuskunze**
**Update:** This PR is now stripped down to straight-forward masking rules. Masking functionality requiring more involved changes of critical code paths are left for follow-up PRs. 

Implements `mask` for `conv`, `pad`, `transpose`, `lax.slice`, `select`, reduce ops.

### CODE REVIEW

**Comment by JuliusKunze** [comment]
Not yet ready for review yet. @mattjj Any idea why CI is not running on new commits?

File: jax/interpreters/masking.py [hunk 1]
```diff
@@ -340,23 +361,74 @@ def lift(self, val):
...
     if not primitive.multiple_results:
       return MaskTracer(self, out, out_shape)
     else:
       return map(partial(MaskTracer, self), out, out_shape)
 
-  def process_call(self, call_primitive, f: lu.WrappedFun, tracers, params):
-    raise NotImplementedError  # TODO mask-of-jit
+  def process_call(self, call_primitive, f, tracers, params):
+    vals, polymorphic_shapes = unzip2((t.val, t.polymorphic_shape) for t in tracers)
+    f, out_shapes_thunk = mask_subtrace(f, self.master, polymorphic_shapes)
```
Comment (j-towns) [inline jax/interpreters/masking.py:387]: We need to be careful about the signature of `mask_subtrace`. The jit cache is keyed on transformations and the 'static' arguments provided to them, in this case `master` and `polymorphic_shapes`, and currently the jitted function will have no way of seeing changes to the logical shapes.

You can see the problem if you do:
```python
In [1]: import jax
   ...: import numpy as np
   ...: import functools

In [2]: @functools.partial(jax.mask, in_shapes=['n'], out_shape='')
   ...: def padded_sum(x):
   ...:   return jax.jit(np.sum)(x)
   ...: 
   ...: print(padded_sum([np.array([3, 1, 4, 1, 5])], dict(n=3)))  # 8
8

In [3]: print(padded_sum([np.array([3, 1, 4, 1, 5])], dict(n=2)))  # Should be 4
8
```
I'm not sure how to fix this right now, just pointing out that it's incorrect 😁.

File: jax/interpreters/masking.py [hunk 2]
```diff
@@ -213,16 +229,16 @@ def is_constant(self):
 
 class Mon(dict):
   def __hash__(self):
-    return hash(tuple(self.items()))
+    return hash(tuple(sorted(self.items())))
```
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/interpreters/masking.py:232]: I notice the same is true of `Poly.__hash__`/`Mon.__lt__` (i.e. they use sorted + lt instead of frozenset(items)), this would make sense if it was possible to get repeated items but I don't think that is possible, so maybe we should change that too? Apologies if I've got this wrong...

File: jax/lax/lax.py [hunk 3]
```diff
@@ -2433,6 +2433,34 @@ def _conv_general_dilated_batch_rule(
       out = _reshape_axis_into(out_spec[1], out_spec[1] + 1, out)
       return out, out_spec[1]
 
+def _masked(x, logical_shape, dimensions, value=0):
```
Comment (j-towns) [inline jax/lax/lax.py:2436]: Could you add a brief comment or docstring explaining what this function does?

File: jax/lax/lax.py [hunk 4]
```diff
@@ -2433,6 +2433,34 @@ def _conv_general_dilated_batch_rule(
...
+  for mask in masks[1:]:
+    mask_intersection &= mask
+  return select(mask_intersection, x, full_like(x, value))
+
+def _conv_general_dilated_masking_rule(
+        padded_vals, logical_shapes, window_strides, padding, lhs_dilation,
+        rhs_dilation, dimension_numbers, feature_group_count, batch_group_count,
+        lhs_shape, rhs_shape, precision):
+  lhs, rhs = padded_vals
+  logical_lhs_shape, rhs_shape = logical_shapes
```
Comment (j-towns) [inline jax/lax/lax.py:2452]: AFAICT `rhs_shape` is ignored. Can you add a check that it isn't polymorphic?

**Review by j-towns — CHANGES_REQUESTED** [review]
Nice work! Some high level/general points:

1. Can you try to keep line lengths to around 80? IMO it's ok to break this rule sometimes but I try to do it as little as possible.
2. mask(jit) is not working correctly, see comment on `process_call`.
3. Perhaps you should say in the pr desription that you've added some support for striding by allowing the user to provide any degree 1 polynomial as a polymorphic shape dimension.

I haven't had time to look over the actual rule implementations yet, so I may follow up with more review later or tomorrow.

File: jax/interpreters/masking.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/interpreters/masking.py:232]: I see now that `sorted` is also used for building string representations, and that makes sense I think, so I guess the thing to do is leave the `__lt__` methods there, but use `frozenset` for hashing.

File: jax/interpreters/masking.py [hunk 5]
```diff
@@ -340,23 +361,74 @@ def lift(self, val):
...
+    return vals, todo
+
+class UniqueId:
+  def __init__(self, name):
+    self.name = name
+
+  def __repr__(self):
+    return self.name
+
+  def __lt__(self, other):
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply jax/interpreters/masking.py:406]: @j-towns This is needed for concise error messages, removing this would break i. e. `masking_test.test_output_shape_error`.

File: jax/interpreters/masking.py [hunk 6]
```diff
@@ -340,23 +361,74 @@ def lift(self, val):
...
+    self[key] = unique_id
+    return unique_id
+
+def remap_ids(names, shape_spec):
+  return ShapeSpec(Poly({Mon({names[id] : deg for id, deg in mon.items()})
+                         : coeff for mon, coeff in poly.items()})
+                   if poly is not _monomorphic_dim else
+                   _monomorphic_dim for poly in shape_spec)
+
+def bind_shapes(polymorphic_shapes, shapes):
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Done!

File: jax/interpreters/masking.py [hunk 7]
```diff
@@ -340,23 +361,74 @@ def lift(self, val):
...
+  env = {}
+  for polymorphic_shape, shape in zip(polymorphic_shapes, shapes):
+    for poly, d in zip(polymorphic_shape, shape):
+      if type(poly) is not Poly or poly.is_constant:
+        if int(poly) != d: raise ShapeError
+      else:
+        poly = poly.copy()
+        const_coeff = poly.pop(Mon({}), 0)
+        ((id,), linear_coeff), = poly.items()
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Done!

File: jax/lax/lax.py (hunk 3 above)
Comment (JuliusKunze) [reply jax/lax/lax.py]: Sure, done!

File: jax/lax/lax.py (hunk 4 above)
Comment (JuliusKunze) [reply jax/lax/lax.py:2456]: Added a check that `rhs_shape` is unpadded, i. e. `logical_shape==padded_shape`. `rhs_shape` can be polymorphic: In the tests it is `(2, 3, i, o)`.

File: jax/api.py [hunk 8]
```diff
@@ -1168,69 +1167,64 @@ def pfun(*args):
...
-               for out, shape in zip(outs, out_shapes)):
-      raise masking.ShapeError
-    return tree_unflatten(out_tree(), outs)
+    flat_fun, out_tree_thunk = flatten_fun_nokwargs(f, in_tree)
+    outs, out_shapes = masking.mask_fun(
+      flat_fun, logical_env, padded_env, args_flat, in_shapes)
+    out_tree = out_tree_thunk()
+    out_shapes = list(out_shapes)
+
+    if out_shape:
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply jax/api.py:1196]: Done!

File: jax/interpreters/masking.py (hunk 1 above)
Comment (j-towns) [reply jax/interpreters/masking.py:389]: I think to get this working we need to move the `extend_shape_envs` call from `masking.mask_fun` into `masking.mask_subtrace`. We also need to
1. Move `padded_env` binding into `mask_subtrace` (from api.mask)
2. Pass `logical_env` (in flattened form) to mask_subtrace as a dynamic argument

I'm wondering whether that should be in a separate pr though?

File: tests/masking_test.py [hunk 9]
```diff
@@ -12,51 +12,63 @@
...
-      ['(-10, m)', 'ShapeSpec(-10, m)'],
-      ['(3 * m)', 'ShapeSpec(3 m)'],
-      ['m', 'ShapeSpec(m)'],
-      ['', 'ShapeSpec()'],
-      ['m + n', 'ShapeSpec(m + n)'],
-      ['m + n * k', 'ShapeSpec(m + k n)'],
-      ['m + 3 * k', 'ShapeSpec(3 k + m)'],
-      ['', 'ShapeSpec()'],
-      ['_', 'ShapeSpec(_)'],
+    ['(m, n)', 'ShapeSpec(m, n)'],
```
Comment (j-towns) [inline tests/masking_test.py:51]: This is a minor nit, but can you revert the indentation on these lines so that they appear the same in the diff?

File: tests/masking_test.py [hunk 10]
```diff
@@ -325,10 +138,102 @@ def padded_sum(x):
     expected = np.array([0, 1, 2, 3, 4])
     self.assertAllClose(ans, expected, check_dtypes=False)
 
+  def check(self, fun, input_shapes, values_dict,
+            out_shape=None, unpadded_vars=None, custom_inputs=None,
+            skip_shapecheck=False, check_output_fun=None):
```
Comment (j-towns) [inline tests/masking_test.py:143]: This helper function looks really bloated. Doesn't it make sense to have separate tests for masking and shape checking? I think having more, fine-grained tests is probably better than few tests where each test checks two different things.

I'm also not sure whether we need to check vmap(mask) for every primitive, since by that logic you should probably test transformation(mask) for every other transformation. I think it makes sense to test vmap(mask) for one or two simple primitives, and any other primitives that we suspect might be problematic. What do you think?

My suggestion would be to revert to the style we had on master, but feel free to disagree and explain why you think this is better. If nothing else it makes the diff a lot easier to review 😜.

**Review by j-towns — CHANGES_REQUESTED** [review]

File: tests/masking_test.py (hunk 10 above)
Comment (j-towns) [reply tests/masking_test.py:143]: On call @JuliusKunze has convinced me that we should test vmap(mask) for each primitive, however it might still be worth trying to simplify this function.

File: jax/interpreters/masking.py (hunk 1 above)
Comment (j-towns) [reply jax/interpreters/masking.py]: Resolved.

File: jax/lax/lax.py [hunk 11]
```diff
@@ -2619,21 +2655,10 @@ def _dot_general_translation_rule(c, lhs, rhs, *, dimension_numbers, precision):
...
-  else:
-    masks = [broadcasted_iota(onp.int32, lhs.shape, d) < lhs_shape[d]
-            for d in lhs_contract]
-    mask_intersection = masks[0]
-    for mask in masks[1:]:
-      mask_intersection &= mask
-    masked_lhs = select(mask_intersection, lhs, zeros_like_array(lhs))
-    return dot_general(masked_lhs, rhs, dimension_numbers, precision=precision)
+  lhs_shape, _ = logical_shapes
+  (lhs_contract, _), _ = dimension_numbers
```
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/lax/lax.py:2662]: Resolved.

File: jax/lax/lax.py [hunk 12]
```diff
@@ -3123,13 +3164,21 @@ def _select_batch_rule(batched_args, batch_dims, **unused_kwargs):
     on_false = broadcast(on_false, pred.shape)
   return select(pred, on_true, on_false), 0
 
+def _select_masking_rule(padded_vals, logical_shapes):
+  pred_shape, true_shape, false_shape = map(lambda x: masking.padded_shape_as_value(x.shape), padded_vals)
+  assert onp.array_equal(pred_shape, true_shape)
+  assert onp.array_equal(pred_shape, false_shape)
+  # ensure all arguments have the same, concrete shape:
+  return select(*(reshape(val, pred_shape) for val in padded_vals))
```
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/lax/lax.py:3175]: It seems the rming the reshape has no effect. I wonder if it only mattered for situations which are now unsupported (particularly since there is now no reshape masking rule).

File: jax/lax/lax.py [hunk 13]
```diff
@@ -3123,13 +3164,21 @@ def _select_batch_rule(batched_args, batch_dims, **unused_kwargs):
     on_false = broadcast(on_false, pred.shape)
   return select(pred, on_true, on_false), 0
 
+def _select_masking_rule(padded_vals, logical_shapes):
+  pred_shape, true_shape, false_shape = map(lambda x: masking.padded_shape_as_value(x.shape), padded_vals)
+  assert onp.array_equal(pred_shape, true_shape)
+  assert onp.array_equal(pred_shape, false_shape)
+  # ensure all arguments have the same, concrete shape:
```
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/lax/lax.py:3174]: Resolved.

File: jax/lax/lax.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/lax/lax.py:3175]: (Resolved)

File: jax/lax/lax.py [hunk 14]
```diff
@@ -3211,10 +3264,19 @@ def _slice_batching_rule(batched_args, batch_dims, *, start_indices,
   out = slice(operand, new_start_indices, new_limit_indices, new_strides)
   return out, bdim
 
+def _slice_masking_rule(padded_vals, logical_shapes, start_indices, limit_indices,
+                        strides):
```
[the remark this answers is not in the document]
Comment (j-towns) [reply jax/lax/lax.py:3272]: Resolved. Not possible because logical shapes can be Jaxpr traced and therefore you can't check their values.

File: tests/masking_test.py (hunk 9 above)
Comment (j-towns) [reply tests/masking_test.py]: Resolved

**Comment by j-towns** [comment]
@googlebot I consent.

**Review by mattjj — APPROVED** [review]
LGTM, great work!

I verified internal tests pass.
